src/decorate.ts

```diff
@@ -0,0 +1,168 @@
+
+		for (const error of errors) {
+			showErrorAt(error.start, error.end, error.value, vscode.DiagnosticSeverity.Error);
+		}
+
+		for (const [key, message] of messageList?.messages) {
+			const hasMetadata = Array.from(messageList.metadata.keys()).filter((literal) => literal.value === ('@' + key.value));
+			let metadata: Metadata | null = null;
+			if (hasMetadata.length > 0) {
+				metadata = messageList.metadata.get(hasMetadata[0])!;
+			}
+
+			decorateKey(key, metadata, messageList.isReference);
+			decorateMessage(message, metadata);
+			decorateMetadata(message, metadata);
+		}
+
+		for (const [key, metadata] of messageList?.metadata) {
+			const hasMessage = Array.from(messageList.messages.keys()).filter((literal) => '@' + literal.value === key.value);
+			if (hasMessage.length === 0) {
+				showErrorAt(key.start, key.end, `Metadata for an undefined key. Add a message key with the name "${key.value.substring(1)}".`, vscode.DiagnosticSeverity.Error);
+			}
+		}
+
+		this.diagnostics.set(editor.document.uri, diagnosticsList);
+		decorationsMap.forEach((value: vscode.Range[], key: vscode.TextEditorDecorationType) => {
+			editor.setDecorations(key, value);
+		});
+
+		function decorateKey(key: Literal, metadata: Metadata | null, isReference: boolean) {
+			if (keyNameRegex.exec(key.value) === null) {
+				showErrorAt(key.start, key.end, `Key "${key.value}" is not a valid message key. The key must start with a letter and contain only letters, numbers, or underscores.`, vscode.DiagnosticSeverity.Error);
+			} else {
+				if (metadata === null && isReference) {
+					showErrorAt(key.start, key.end, `The message with key "${key.value}" does not have metadata defined.`, vscode.DiagnosticSeverity.Information);
+				}
+			}
+		}
+
+		function decorateMessage(message: Message, metadata: Metadata | null) {
+			if (message instanceof CombinedMessage) {
+				for (const submessage of message.parts) {
+					decorateMessage(submessage, metadata);
+				}
+			} else if (message instanceof Placeholder) {
+				decorateAndValidatePlaceholder(message.placeholder, metadata);
+			} else if (message instanceof ComplexMessage) {
+
+				decorateAndValidatePlaceholder(message.argument, metadata);
+				decorateAt(message.argument.start, message.argument.end, placeholderDecoration);
+				if (placeholderNameRegex.exec(message.argument.value) === null) {
+					showErrorAt(message.argument.start, message.argument.end, `"${message.argument.value}" is not a valid placeholder name. The key must start with a letter and contain only letters, numbers, underscores.`, vscode.DiagnosticSeverity.Error);
+				}
+
+				if (!Array.from(message.messages.keys()).some((p) => p.value.trim() === 'other')) {
+					showErrorAt(message.start + 1, message.end + 1, `The ICU message format requires a 'other' argument.`, vscode.DiagnosticSeverity.Error);
+				}
+
+				if (!['plural', 'select', 'gender'].includes(message.complexType.value.trim())) {
```

Comment:
"gender" seems odd here - is this supposed to be "choice"?

Reply:
It's actually gender - this is a special case for `select` defined in the ICU message format.